lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -665,14 +688,159 @@ def write(chunk)
+                increment_dropped_entries_count(partial_error_count)
+                @log.warn "Dropping #{partial_error_count} log message(s)",
+                          error_code: "google.rpc.Code[#{error_code}]",
+                          error: error_message
+              end
+            end
           end
         end
       end
     end
 
+    # Extract a map of error details from an potentially partially successful
+    # request. The keys in this map are stringified indexes of the failed log
+    # entries in the original json payload, and the values are the error code
+    # and error message.
+    #
+    # A sample error.body looks like:
+    # {
+    #   "error": {
+    #     "code": 403,
+    #     "message": "User not authorized.",
+    #     "status": "PERMISSION_DENIED",
+    #     "details": [
+    #       {
+    #         "@type": "type.googleapis.com/google.logging.v2.WriteLogEntriesPar
+    #           tialErrors",
+    #         "logEntryErrors": {
+    #           "0": {
+    #             "code": 7,
+    #             "message": "User not authorized."
+    #           },
+    #           "1": {
+    #             "code": 3,
+    #             "message": "Log name contains illegal character :"
+    #           },
+    #           "3": {
+    #             "code": 3,
+    #             "message": "Log name contains illegal character :"
+    #           }
+    #         }
+    #       },
+    #       {
+    #         "@type": "type.googleapis.com/google.rpc.DebugInfo",
+    #         "detail": ...
+    #       }
+    #     ]
+    #   }
+    # }
+    #
+    # The root level "code", "message", and "status" simply match the root
+    # cause of the first failed log entry. For example, if we switched the order
+    # of the log entries, then we would get:
+    # {
+    #    "error" : {
+    #       "code" : 400,
+    #       "message" : "Log name contains illegal character :",
+    #       "status" : "INVALID_ARGUMENT",
+    #       "details": ...
+    #    }
+    # }
+    # We will ignore it anyway and look at the details instead which includes
+    # info for all failed log entries.
+    #
+    # In this example, the logEntryErrors we are extracting should be:
+    # {
+    #   "0": {
+    #     "code": 7,
+    #     "message": "User not authorized."
+    #   },
+    #   "1": {
+    #     "code": 3,
+    #     "message": "Log name contains illegal character :"
+    #   },
+    #   "3": {
+    #     "code": 3,
+    #     "message": "Log name contains illegal character :"
+    #   }
+    # }
+    def extract_log_entry_errors(error)
+      error_body = ensure_hash(JSON.parse(error.body))
+      error = ensure_hash(error_body['error'])
+      error_details = ensure_array(error['details'])
+      partial_errors = error_details.detect do |error_detail|
+        ensure_hash(error_detail)['@type'] == PARTIAL_ERROR_FIELD
+      end
+      unless partial_errors
+        @log.warn 'Failed to detect an error detail with type' \
```

Comment:
This will warn twice -- once here, and once when catching the exception from the later `ensure_hash`.
My suggestion is to throw the exception from the _ifnone_ function of `Array#detect`:
```ruby
partial_errors = ensure_array(error['details']).detect(
  lambda { fail JSON::ParserError.new("No type #{PARTIAL_ERROR_FIELD}") }
) do |error_detail|
  ensure_hash(error_detail)['@type'] == PARTIAL_ERROR_FIELD
end
```

Reply:
Changed.